### ISSUE TITLE
Invalid dune file in json-data CI job

### ISSUE THREAD

**Issue by palmskog** [coq/opam-coq-archive#1995]
We recently started seeing failing `json-data` CI jobs, with the following message:
```
File "nix/store/2w9yyr6c3py70kwsv8zvvx5afsfx3zx3-dune-2.8.5/bin/dune", line 1, characters 0-0:
Error: Invalid dune file
Cleaning up project directory and file based variables
ERROR: Job failed: exit code 1
```
As was figured out by @silene in #1994, the error message only appears with "latest" Nix Docker images, and disappears with `nixos/nix:2.3.12`. However, setting the latter may only be a band-aid for the underlying issue, which may be some bad interaction between Nix and Dune.

Example of failing job: https://gitlab.com/coq/opam-coq-archive/-/jobs/1889653040

Pinging Nix-knowledgeable people who may be able to comment: @Zimmi48 @vbgl.

**Comment by Zimmi48** [comment]
I don't know much about how these Docker images work, but from the version names, I would expect that they are linked to the underlying version of Nix (Nix 2.3 has been the current version for a very long time, and 2.5 was out just three days ago). The error might also be related to the nixpkgs version. I couldn't reproduce with `nix-shell --run "dune exec --profile=release -- archive2web released extra-dev > coq-packages.json"` on my machine, but I am still on Nix 2.3.15.

**Comment by palmskog** [comment]
If 2.5 appeared just three days ago, then this coincides with the beginning of our failing jobs here quite closely. But someone should try on 2.5 the command tried by Théo: `nix-shell --run "dune exec --profile=release -- archive2web released extra-dev > coq-packages.json"`

**Comment by Zimmi48** [comment]
What I find problematic for reproducibility is that `nixpkgs` is not pinned in `default.nix`. The error could very well be related, not to the Nix version itself, but to the nixpkgs version. And the Docker image might just never get an updated nixpkgs, which would mean that it got updated only when the Nix version got updated.

**Comment by Zimmi48** [comment]
I tried to update just the version of nixpkgs (but not the Nix version) to the latest `nixpks-unstable`, but still couldn't reproduce the issue.

**Comment by vbgl** [comment]
Can’t reproduce.

I don’t know how to use `docker`. Can someone run `nix-info` (or extract useful information in any way) in the docker container mentioned in the failing build log?
> Using docker image sha256:118288c8ceba03c94fae86b5dd0a8e0b7e7c473971fe2d824efce7f135740442 for nixos/nix with digest nixos/nix@sha256:90adfdfd7caf8bd33b055e3b81b69104e16d251abc87b8d8221ed71a5a769846

Thanks.

**Comment by gares** [comment]
Do we really need nix to compile 1 ocaml file?
I mean, the other jobs already have a cache with ocaml.

I like nix, but this time we are just looking for extra trouble.

**Comment by Zimmi48** [comment]
I indeed suggest removing the dependency on Nix. If we want to use Nix, we should use proper reproducibility best practices. If we don't, then I don't see the point.

**Comment by Zimmi48** [comment]
For the record, I tried to reproduce with Docker, but then I get errors of the type:

```
unexpected error in download thread: error: unsharing filesystem state in download thread: Operation not permitted
warning: download of 'https://cache.nixos.org/nix-cache-info' was interrupted
[...]
building '/nix/store/hd3ajzairvilig9hqgwqsj1s34q01n1w-binutils-2.35.1.tar.bz2.drv'...
unexpected error in download thread: error: unsharing filesystem state in download thread: Operation not permitted
error: download of 'https://ftpmirror.gnu.org/binutils/binutils-2.35.1.tar.bz2' was interrupted
error: builder for '/nix/store/hd3ajzairvilig9hqgwqsj1s34q01n1w-binutils-2.35.1.tar.bz2.drv' failed with exit code 1
```

so I'm giving up on reproducing this issue.

**Comment by palmskog** [comment]
OK, then I guess the consensus is to remove the Nix job, and I welcome any PR doing that.

However, I think we can now say with some confidence that the failures are due to Nix Docker images rolling out 2.5 as their `latest` Docker tag (https://hub.docker.com/r/nixos/nix/tags), so at least we should be good for a while on the old 2.3 - the constant red CI was becoming a big problem when reviewing PRs, but was confirmed to disappear in #1994.

**Comment by silene** [comment]
To be clear, the failure "unsharing filesystem state in download thread: Operation not permitted" is widely reported on NixOS 2.5 and it is what prompted me to try 2.3.12 instead.

As for using Docker, it is not too difficult, just type `docker run -ti --mount type=bind,source=$PWD,target=/foo nixos/nix:2.3.12`. Assuming `PWD` was the Opam Coq repository, you can then `cd /foo` and run `nix-shell --run "dune exec --profile=release -- archive2web released extra-dev > coq-packages.json"`.

As for `nix-info`, it is not part of the `nixos/nix` image, so no idea how to run it (or even install it).

**Comment by vbgl** [comment]
Thanks for your reply.

> As for `nix-info`, it is not part of the `nixos/nix` image, so no idea how to run it (or even install it).

It is not too difficult, just type `nix-shell -p nix-info --run nix-info`.

**Comment by silene** [comment]
On the 2.5 image, it fails with the bug already mentioned: `unsharing filesystem state in download thread: Operation not permitted`

On the 2.3.12 image; it gives:
```console
# nix-shell -p nix-info --run nix-info
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.3.12, channels(root): "nixpkgs-21.11pre295944.0747387223e", nixpkgs: /nix/var/nix/profiles/per-user/root/channels/nixpkgs
```

**Comment by Zimmi48** [comment]
The surprising thing is that `nix-shell` doesn't fail in the same way when run in CI...

**Comment by gares** [comment]
> I welcome any PR doing that.

While you guy debug this nix mistery, I'll work on this PR

**Comment by silene** [comment]
To be more precise, it does not fail in the same way when run in CI *on Gitlab's shared runners*! Because if I run it on my computer using `gitlab-runner exec docker json-data`, I do get `unsharing filesystem state in download thread: Operation not permitted`. So, the shared runners also have an impact here. If this is due to a data race, the slowness of the runners might play a significant role.

**Comment by gares** [comment]
Just for fun, nix is not guilty: https://gitlab.com/coq/opam-coq-archive/-/jobs/1892693440
I will build using findlib...

**Comment by ejgallego** [comment]
Where's the code that's calling `dune exec` ?

**Comment by gares** [comment]
here: https://github.com/coq/opam-coq-archive/blob/master/.gitlab-ci.yml#L158

**Comment by Zimmi48** [comment]
> here: https://github.com/coq/opam-coq-archive/blob/master/.gitlab-ci.yml#L158

Your link is not a permalink, so it's already outdated.

**Comment by ejgallego** [comment]
As noted in Zulip you need to either `cd` to the right directory where the workspace is, or use `--root .`.

Calling `dune exec` not from a dune workspace will be a direct failure in 3.0 , as opposed to now which tries to detect the workspace, and if not present, assumes there is one in the cwd.

**Comment by silene** [comment]
According to Dune's documentation, "the root of the current workspace is determined by looking up a [...] `dune-project` file in the current directory [...]." And there is a `dune-project` in the directory from which `dune exec` is called.

Or is the documentation wrong? How many files do we need to put in the current directory before we can even start to use Dune? `dune-workspace`, `dune-project`, and `dune`?

**Comment by ejgallego** [comment]
It is called in the immediate ancestor of the one with `dune-project`

**Comment by silene** [comment]
How do you know that?

The file `coq-packages.json` is created in the correct directory (the one containing `dune-project`). So, how could `dune exec ... > coq-packages.json` have been executed in the wrong directory?

**Comment by SkySkimmer** [comment]
The way the error message says `File "nix/store/..."` without a leading / implies it's executed in `/`. That would explain why it looks in such a path for dune files too.

**Comment by silene** [comment]
No, it does not imply that Dune was started from `/`. It only implies that Dune, at some point, did `Sus.chdir "/" (which it does, according to `strace`). And since Dune is also using threads, I wonder what could go wrong...
